mopidy_tidal/login_hack.py

```diff
@@ -0,0 +1,304 @@
+                k: next(iter(args[i : i + 1]), None)
+                for i, k in enumerate(spec_args[1:])
+            }
+            if "uri" in repr(spec):
+                uri = find_uri(args_mapping, kwargs)
+                if uri:
+                    _, schema, *_ = uri.split(":")
+            elif "field" in spec.args:
+                schema = kwargs.get("field", args_mapping["field"])
+
+            if schema:
+                type_mapping = {
+                    "artists": Artist,
+                    "albums": Album,
+                    "playlists": Playlist,
+                    "tracks": Track,
+                    "moods": Ref,
+                    "mixes": Ref,
+                    "genres": Ref,
+                }
+                plural = (
+                    schema
+                    if schema.endswith("s")
+                    else schema + ("es" if schema.endswith("x") else "s")
+                ).replace("my_", "")
+                expected_return_type = type_mapping.get(plural, NoneType)
+
+            if manual_return_type:
+                return_type = manual_return_type
+            else:
+                # Assume all declared return types have the same structure
+                declared_return_types = extract_types(fn.__annotations__["return"])
+                possible_return_types = reduce(
+                    lambda a, b: a | interesting_types(b),
+                    [set(), *declared_return_types],
+                )
+                expected_type_possible = (
+                    expected_return_type is not NoneType
+                    and interesting_types(expected_return_type).issubset(
+                        possible_return_types
+                    )
+                )
+                return_type = (
+                    match_structure(declared_return_types[0], expected_return_type)
+                    if expected_type_possible
+                    else declared_return_types[0]
+                )
+
+            url = backend.login_url
+            msg = f"Please visit {url} to log in."
+
+            if passthrough:
+                return PassthroughBuilder(by_type={str: lambda: msg}).build(return_type)
+            else:
+                return ObjectBuilder(schema=schema, uri=uri, url=url, msg=msg).build(
+                    return_type
+                )
+        elif backend.login_method == "HACK":
+            audio_helper = LoginAudioHelper(backend)
+            audio_helper.remove()
+
+        return fn(obj, *args, **kwargs)
+
+    return wrapper
+
+
+def match_structure(target_type, inner_type):
+    """Return a type struture equivalent to the target but with the right inner type."""
+    base_type = get_origin(target_type)
+    if base_type is dict:
+        return base_type[get_args(target_type)[0], inner_type]
+    if base_type is list:
+        return base_type[inner_type]
+    else:
+        return inner_type
+
+
+voice_rss_api_key = "eb909fe9f2ce403bb7209de172d096f1"
```

Comment:
Do we want the API key directly in the sources or should we add it as a secret?

Reply:
Thing is it has to be used by the *instance*, so we have to ship it.  I deliberately picked a service which didn't require a credit card and has a free tier, so I'm not that worried about exposing it.  If you like I could obfuscate it a bit (e.g. use base64) to defend against automated scraping.  I couldn't think of a good solution though and did try to find a service which didn't require an api key.